include/libtorrent/kademlia/dht_storage.hpp

```diff
@@ -0,0 +1,216 @@
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+/*
+
+Second author: Alden Torres (aldenml)
+Same copyright
+
+*/
+
+#ifndef DHT_STORAGE_HPP
+#define DHT_STORAGE_HPP
+
+#include <algorithm>
+#include <map>
+#include <set>
+#include <string>
+
+#include <libtorrent/config.hpp>
+#include <libtorrent/time.hpp>
+#include <libtorrent/socket.hpp>
+#include <libtorrent/sha1_hash.hpp>
+#include <libtorrent/bloom_filter.hpp>
+#include <libtorrent/address.hpp>
+#include <libtorrent/session_settings.hpp>
+#include <libtorrent/performance_counters.hpp>
+
+#include <libtorrent/kademlia/item.hpp>
+#include <libtorrent/kademlia/node_id.hpp>
+
+namespace libtorrent { namespace dht
+{
+	// this is the entry for every peer
+	// the timestamp is there to make it possible
+	// to remove stale peers
+	struct peer_entry
+	{
+		time_point added;
+		tcp::endpoint addr;
+		bool seed;
+	};
+
+	// this is a group. It contains a set of group members
+	struct torrent_entry
+	{
+		std::string name;
+		std::set<peer_entry> peers;
+	};
+
+	struct dht_immutable_item
+	{
+		dht_immutable_item() : value(0), size(0) {}
+		// malloced space for the actual value
+		char* value;
+		// size of malloced space pointed to by value
+		int size;
+	};
+
+	struct ed25519_public_key { char bytes[item_pk_len]; };
+
+	struct dht_mutable_item : dht_immutable_item
+	{
+		char sig[item_sig_len];
+		boost::uint64_t seq;
+		ed25519_public_key key;
+		char* salt;
+		int salt_size;
+	};
+
+	struct TORRENT_EXPORT dht_storage_interface
+	{
+#ifndef TORRENT_NO_DEPRECATE
+		virtual size_t num_torrents() const = 0;
+		virtual size_t num_peers() const = 0;
+#endif
+
+		// returns true if the entry is found and it is returned in the
+		// out parameter entry
+		virtual bool lookup_peers(sha1_hash const& info_hash
+			, torrent_entry& entry) const = 0;
+		// this method is named announce_peer for consistency with the
+		// upper layers, but has nothing to do with networking. It's only
+		// responsibility is store the peer in such a way that it's returned
+		// in the torrent_entry with the lookup_peers.
+		virtual void announce_peer(sha1_hash const& info_hash
+			, tcp::endpoint const& endp
+			, std::string const& name, bool seed) = 0;
+
+		// returns true if the item is found and it is returned in the
+		// out parameter item
+		virtual bool get_immutable_item(sha1_hash const& target
+			, dht_immutable_item& item) const = 0;
+		virtual void put_immutable_item(sha1_hash const& target
+			, dht_immutable_item const& to_add, address addr) = 0;
+
+		// returns true if the item is found and it is returned in the
+		// out parameter item
+		virtual bool get_mutable_item(sha1_hash const& target
+			, dht_mutable_item& item) const = 0;
+		virtual void put_mutable_item(sha1_hash const& target
+			, dht_mutable_item const& to_add, address addr) = 0;
+
+		virtual void tick() = 0;
+	};
+
+	// NON-PUBLIC TYPES
```

Comment:
ideally they can be moved to the .cpp file (in an anonymous namespace), or at least some of them.